src/components/page-details.jsx

```diff
@@ -176,15 +170,34 @@ export default class PageDetails extends React.Component {
 
   _versionsToRender () {
     const [fromId, toId] = (this.props.match.params.change || '').split('..');
-    let from = this.state.page.versions.find(v => v.uuid === fromId);
-    let to = this.state.page.versions.find(v => v.uuid === toId);
+    const versions = this.state.page.versions;
+    let from, to, shouldRedirect = false;
 
-    // Changes with no `to` are invalid, but those where `from` was never
-    // specified are OK (it’ll be considered relative to `to`)
-    if (!to || fromId && !from) {
-      to = from = null;
+    from = versions.find(v => v.uuid === fromId);
+    to = versions.find(v => v.uuid === toId);
+
+    if (!(from && to)) {
+      shouldRedirect = true;
+
+      if (toId === '$') {
+        to = versions[0];
+      }
+      else if (from) {
+        to = versions[versions.findIndex(v => v.uuid === from.uuid) - 1];
+      }
+      else {
+        to = to || versions[0];
+      }
+
+      if (fromId === '^') {
+        from = versions[versions.length - 1];
+      }
+      else {
+        from = from || versions.find(v => v.capture_time < to.capture_time) || to;
+      }
```

Comment:
I think moving all the lookup & decisions about redirects, etc. here makes sense. 👍 

This definitely works. If you want to go with this, great. 👌  That said, I do think this logic is a little hard to follow (and your concern about being able to explain it is maybe also a good indicator of that). It might be clearer to break it down more explicitly, e.g.

```js
// Look up explicit version IDs
if (toId === '$') {
  to = versions[0];
  shouldRedirect = true;
}
else if (toId) {
  to = versions.find(v => v.uuid === toId);
  if (!to) {
    throw new Error(`There is no version “${toId}”`);
  }
}

// ...repeat for fromId

// Handle empty string version IDs (they are relative to the other version)
if (!from && !to) {
  throw new Error('No versions specified');
}
else if (from && !to) {
  to = versions[versions.indexOf(from) - 1] || from;
  shouldRedirect = true;
}
else {
  from = versions[versions.indexOf(to) + 1] || to;
  shouldRedirect = true;
}
```

Or even abstract out the lookup part:

```js
function findVersion (id, versions) {
  if (id === '^') {
    // at the end, check id vs. the object’s uuid to determine whether to redirect:
    //  shouldRedirect = from.uuid === fromId && to.uuid === toId
    return versions[versions.length - 1];
    // or you could use some other structure to indicate this is a “pointer”
    //   {uuid: versions[versions.length - 1].uuid, redirect: true};
    //   {redirect: versions[versions.length - 1]}
  }
  if (id === '$') {
    return versions[0];
  }

  const version = versions.find(v => v.uuid === id);
  if (version) {
    return version;
  }

  throw new Error(`There is no version “${id}”`);
}

let to = toId && findVersion(toId, versions);
let from = fromId && findVersion(fromId, versions);
// now handle relative IDs
```

Reply:
I tried both methods, but I couldn't get them to cover all the different cases without becoming very complicated or not working right for me. 

For the first suggestion if we throw an error when `if (!to)`, then we wouldn't be able to check for relative paths later with `if (!from && !to)`. I also found abstracting out the lookup a little harder for me to follow.

I was able to simplify the `else` clauses I was using into single-line, consistent short-circuit evaluations and added comments. Hopefully that makes the logic easier to follow. 

Everything still works, so I'll go ahead and merge.

Reply:
> For the first suggestion if we throw an error when `if (!to)`, then we wouldn't be able to check for relative paths later with if `(!from && !to)`.

We would, because that error is only thrown if `toId` was specified. i.e. if you specify an ID that doesn’t work out to an actual version, we shouldn’t redirect to some other version; we should tell you that version doesn’t exist. The point was that we should *not* move on to choosing a new version in that case.

Reply:
I think I understand what you're saying now and must have done something wrong before, but will take a look at this again in the next couple days.

TBH, every variation makes my brain hurt a bit. Is this a normal thing to do with shortcuts in routing? 😵